feat: refactor and add protocol support to `stats/base/nanvariancewd`

prajjwalbajpai opened the pull request:
Resolves #5677.

## Description

> What is the purpose of this pull request?

This pull request:

-   Adds support for accessor arrays and refactor stats/base/nanvariancewd

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #5677

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

prajjwalbajpai committed:
feat: add support for accessor arrays and refactor stats/base/nanvariancewd

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanvariancewd/README.md:
```diff
@@ -114,17 +114,14 @@ The function has the following parameters:
-var N = floor( x.length / 2 );
 
-var v = nanvariancewd( N, 1, x, 2 );
+var v = nanvariancewd( 4, 1, x, 2 );
```
```suggestion
var v = nanvariancewd( 5, 1, x, 2 );
```
you should explicitly traverse over the `NaN` element to show how we're handling it

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanvariancewd/README.md:
````diff
@@ -158,17 +152,14 @@ var v = nanvariancewd.ndarray( x.length, 1, x, 1, 0 );
 ```javascript
-var floor = require( '@stdlib/math/base/special/floor' );
-
 var x = [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 ];
````
```suggestion
var x = [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0, NaN, NaN ];
```

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanvariancewd/README.md:
```diff
@@ -158,17 +152,14 @@ var v = nanvariancewd.ndarray( x.length, 1, x, 1, 0 );
-var N = floor( x.length / 2 );
 
-var v = nanvariancewd.ndarray( N, 1, x, 2, 1 );
+var v = nanvariancewd.ndarray( 4, 1, x, 2, 1 );
```
```suggestion
var v = nanvariancewd.ndarray( 5, 1, x, 2, 1 );
```
same comment regarding documenting the handling of `NaN` elements

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanvariancewd/README.md:
```diff
@@ -195,18 +187,12 @@ var v = nanvariancewd.ndarray( N, 1, x, 2, 1 );
-}
+var x = discreteUniform( 10, -50, 50, {
+    'dtype': 'float64'
+});
```
https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/stats/base/dnanvariancepn/examples/index.js

refer to this as an example

aayush0325 replied:
same comment for `examples/index.js`

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanvariancewd/benchmark/benchmark.js:
https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/stats/base/dnanvariancepn/benchmark/benchmark.js

kindly refer to this as an example

aayush0325 replied:
same comment for your other benchmark file

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanvariancewd/docs/repl.txt:
```diff
@@ -113,8 +110,7 @@
     > var x = [ 1.0, -2.0, 3.0, 2.0, 5.0, -1.0 ];
-    > var N = {{alias:@stdlib/math/base/special/floor}}( x.length / 2 );
-    > {{alias}}.ndarray( N, 1, x, 2, 1 )
+    > {{alias}}.ndarray( 3, 1, x, 2, 1 )
```
same comment regarding documentation of handling of `NaN` elements

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanvariancewd/test/test.main.js:
```diff
@@ -1,3 +1,4 @@
+
```
```suggestion
```

aayush0325 reviewed (changes requested):
kindly address these comments for any other PRs that you may have

prajjwalbajpai committed:
fix: added NaN to test

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

prajjwalbajpai committed:
Merge branch 'stdlib-js:develop' into refactor-nanvariancewd

prajjwalbajpai commented:
@aayush0325 please review the changes

Planeshifter committed:
docs: add returns annotations

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter committed:
chore: add Git note

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter committed:
chore: add Git note

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter committed:
chore: add Git note

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

prajjwalbajpai committed:
fix: fixed errors in examples

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

prajjwalbajpai committed:
Merge branch 'develop' into refactor-nanvariancewd

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanvariancewd/docs/repl.txt:
```diff
@@ -1,10 +1,10 @@
 
-    The `N` and `stride` parameters determine which elements in `x` are accessed
-    at runtime.
+    The `N` and `stride` parameters determine which elements
```
update line wrapping

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanvariancewd/test/test.main.js:
```diff
@@ -94,6 +94,60 @@ tape( 'the function calculates the population variance of a strided array (ignor
 	t.end();
 });
 
+tape( 'the function calculates the population variance of a strided array (ignoring `NaN` values) (accessors)', function test( t ) {
```
```suggestion
tape( 'the function calculates the population variance of a strided array (accessors)', function test( t ) {
```

prajjwalbajpai replied:
Do I have to remove `(ignoring 'NaN' values)` from all tests?
And is this thing something that needs to be done for all nan functions?

aayush0325 committed:
chore: update line wrapping in repl.txt

aayush0325 commented:
@prajjwalbajpai kindly see my latest commit to get an idea of how we want the line wrapping in `repl.txt` to look like, note that the maximum character limit is 80 for a line.

aayush0325 committed:
chore: fix linting error

kgryte commented on lib/node_modules/@stdlib/stats/base/nanvariancewd/docs/repl.txt:
```diff
@@ -1,10 +1,10 @@
 
-    The `N` and `stride` parameters determine which elements in `x` are accessed
-    at runtime.
+    The `N` and `stride` parameters determine which elements in the strided
+    array are accessed at runtime.
```
```suggestion
    The `N` and stride parameters determine which elements in the strided array
    are accessed at runtime.
```

kgryte commented on lib/node_modules/@stdlib/stats/base/nanvariancewd/lib/main.js:
```diff
@@ -20,14 +20,31 @@
+*
+* @example
+* var x = [ 1.0, -2.0, NaN, 2.0 ];
+* var N = x.length;
+*
+* var v = nanvariancewd( N, 1, x, 1 );
```
```suggestion
*
* var v = nanvariancewd( x.length, 1, x, 1 );
```

kgryte commented on lib/node_modules/@stdlib/stats/base/nanvariancewd/README.md:
```diff
@@ -98,7 +98,7 @@ The use of the term `n-1` is commonly referred to as Bessel's correction. Note,
-#### nanvariancewd( N, correction, x, stride )
+#### nanvariancewd( N, correction, x, strideX )
 
 Computes the [variance][variance] of a strided array `x` ignoring `NaN` values and using Welford's algorithm.
```
```suggestion
Computes the [variance][variance] of a strided array ignoring `NaN` values and using Welford's algorithm.
```

kgryte committed:
docs: update copy

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
docs: fix wrapping

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/nanvariancewd/lib/accessors.js:
```diff
@@ -50,33 +59,33 @@ function nanvariancewd( N, correction, x, stride ) {
-	}
-	if ( N === 1 || stride === 0 ) {
-		v = x[ 0 ];
+	// Cache reference to array data:
```
```suggestion
	// Cache a reference to array data:
```

kgryte commented on lib/node_modules/@stdlib/stats/base/nanvariancewd/lib/accessors.js:
```diff
@@ -50,33 +59,33 @@ function nanvariancewd( N, correction, x, stride ) {
-	} else {
-		ix = 0;
-	}
+
```
```suggestion
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
test: add missing tests and update descriptions

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
test: add missing tests and update descriptions

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM after clean-up.
